freqtrade/optimize/hyperopt.py

```diff
@@ -410,6 +410,35 @@ def prepare_hyperopt_data(self) -> None:
         # Store non-trimmed data - will be trimmed after signal generation.
         dump(preprocessed, self.data_pickle_file)
 
+    def get_asked_points(self, n_points: int) -> List[List[Any]]:
+        '''
+        Enforce points returned from `self.opt.ask` have not been already evaluated
+
+        Steps:
+        1. Try to get points using `self.opt.ask` first
+        2. Discard the points that have already been evaluated
+        3. Retry using `self.opt.ask` up to 3 times
+        4. If still some points are missing in respect to `n_points`, random sample some points
+        5. Repeat until at least `n_points` points in the `asked_non_tried` list
+        6. Return a list with length truncated at `n_points`
+        '''
+        i = 0
+        asked_non_tried: List[List[Any]] = []
+        while i < 100 and len(asked_non_tried) < n_points:
```

Comment:
Unfortunately this shows some very bad behavior once a space is "really" exhausted.

I've stress-tested this with 100 epochs and a space of 40 possible points (basically one int space). 
This means starting from epoch 40, it was impossible to get new points.

This led to endless loops of "new random points" (~100 per iteration) - which led to hyperopt taking quite a long time (a considerably longer than on develop) - which is the opposite from what we'd hope.

maybe it's just that 100 cycles/attempts are too much (considering that we'll be getting `3 x "real new points" + 97 * points * 5` - there MUST be some random points available a lot earlier.

Reply:
i'll need to do more tests in different scenarios - but as you'll know yourself - testing things in hyperopt is quite timeconsuming.

the above example (way too small a space of 40 points) shows as follows:

* ~1m for 100 epochs on current develop
* ~5m for 100 epochs on this branch

Now it's obviously a edge-case - one which actually should implement early stopping due to space exhaustion (which is something that scikit-optimize doesn't support afaik).

We might be in a position to do so with this method - but i'm not sure on the logic for that. After how many consecutive _random duplicate_ points would we start to consider a space 100% exhausted?


a more interesting test will be to test it with a bigger space - which i hope to find time for over the weekend.